share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -17,58 +17,565 @@
+        return new CoordinatorResult<>(Collections.emptyList(), responseData);
+    }
+
+    // Visible for testing
+    Integer getLeaderMapValue(SharePartitionKey key) {
+        return this.leaderEpochMap.get(key);
+    }
+
+    // Visible for testing
+    Integer getStateEpochMapValue(SharePartitionKey key) {
+        return this.stateEpochMap.get(key);
+    }
+
+    // Visible for testing
+    ShareGroupOffset getShareStateMapValue(SharePartitionKey key) {
+        return this.shareStateMap.get(key);
+    }
+
+    // Visible for testing
+    CoordinatorMetricsShard getMetricsShard() {
+        return metricsShard;
+    }
+
+    private static ShareGroupOffset merge(ShareGroupOffset soFar, ShareUpdateValue newData) {
+        // snapshot epoch should be same as last share snapshot
+        // state epoch is not present
+        Set<PersisterOffsetsStateBatch> currentBatches = soFar.stateBatchAsSet();
+        long newStartOffset = newData.startOffset() == -1 ? soFar.startOffset() : newData.startOffset();
+        int newLeaderEpoch = newData.leaderEpoch() == -1 ? soFar.leaderEpoch() : newData.leaderEpoch();
+
+        return new ShareGroupOffset.Builder()
+            .setSnapshotEpoch(soFar.snapshotEpoch())
+            .setStateEpoch(soFar.stateEpoch())
+            .setStartOffset(newStartOffset)
+            .setLeaderEpoch(newLeaderEpoch)
+            .setStateBatches(combineStateBatches(currentBatches, newData.stateBatches().stream()
+                .map(PersisterOffsetsStateBatch::from)
+                .collect(Collectors.toCollection(LinkedHashSet::new)), newStartOffset))
+            .build();
+    }
+
+    /**
+     * Util method which takes in 2 collections containing {@link PersisterOffsetsStateBatch}
+     * and the startOffset.
+     * It removes all batches from the 1st collection which have the same first and last offset
+     * as the batches in 2nd collection. It then creates a final list of batches which contains the
+     * former result and all the batches in the 2nd collection. In set notation (A - B) U B (we prefer batches in B
+     * which have same first and last offset in A).
+     * Finally, it removes any batches where the lastOffset < startOffset, if the startOffset > -1.
+     * @param currentBatch - collection containing current soft state of batches
+     * @param newBatch - collection containing batches in incoming request
+     * @param startOffset - startOffset to consider when removing old batches.
+     * @return List containing combined batches
+     */
+    private static List<PersisterOffsetsStateBatch> combineStateBatches(
+        Collection<PersisterOffsetsStateBatch> currentBatch,
+        Collection<PersisterOffsetsStateBatch> newBatch,
+        long startOffset
+    ) {
+        currentBatch.removeAll(newBatch);
+        List<PersisterOffsetsStateBatch> batchesToAdd = new LinkedList<>(currentBatch);
+        batchesToAdd.addAll(newBatch);
```

Comment:
It would be useful to verify if this is an issue. Suppose currentBatch is
batch1 {
firstOffset: 100
lastOffset: 109
deliverState: Available
deliverCount: 1
}
and newBatch is 
batch2{
firstOffset: 105
lastOffset: 105
deliverState: Acknowledge
deliverCount: 1
}

After the call to combineStateBatches(), Share coordinator will have both batches in its state and thus the share leader could have both batches too (after initializing from ReadShareGroupState). Now suppose that the share leader fetches the following batch and calls SharePartition.acquire().

fetchedBatch{
firstOffset: 100
lastOffset: 109
}

Both batch1 and batch2 will match the fetched batch. When calling acquireSubsetBatchRecords() on batch1, we will add the full batch to AcquiredRecords. When calling acquireSubsetBatchRecords() on batch2, we will skip since the only record in it has been acked. But AcquiredRecords is unchanged after this. This means that we will return the full batch as acquired records, which is incorrect since offset 105 shouldn't be acquired.